{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a5ad9d33-abd2-4c9e-b7ce-fa57c2ffb522",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "\"\"\"\n",
    "This is a python script that analyse Amplicon FASTQ file generated with Illumina Sequencing\n",
    "The script classify which amino acids are present at a certain position\n",
    "The purpose is to see which amino acid mutations have been generated\n",
    "All the data are exported into a csv\n",
    "\"\"\"\n",
    "\n",
    "# Translate a codon of 3 bases into the appropriate amino acids\n",
    "def amino_acids(aa):\n",
    "    if aa == \"TTT\" or aa == \"TTC\":\n",
    "        return \"F\"\n",
    "    if aa == \"TTA\" or aa == \"TTG\" or aa == \"CTT\" or aa == \"CTC\" or aa == \"CTA\" or aa == \"CTG\":\n",
    "        return \"L\"\n",
    "    if aa == \"ATT\" or aa == \"ATC\" or aa == \"ATA\":\n",
    "        return \"I\"\n",
    "    if aa == \"ATG\":\n",
    "        return \"M\"\n",
    "    if aa == \"GTT\" or aa == \"GTC\" or aa == \"GTA\" or aa == \"GTG\":\n",
    "        return \"V\"\n",
    "    if aa == \"TCT\" or aa == \"TCC\" or aa == \"TCA\" or aa == \"TCG\" or aa == \"AGT\" or aa == \"AGC\":\n",
    "        return \"S\"\n",
    "    if aa == \"CCT\" or aa == \"CCC\" or aa == \"CCA\" or aa == \"CCG\":\n",
    "        return \"P\"\n",
    "    if aa == \"ACT\" or aa == \"ACC\" or aa == \"ACA\" or aa == \"ACG\":\n",
    "        return \"T\"\n",
    "    if aa == \"GCT\" or aa == \"GCC\" or aa == \"GCA\" or aa == \"GCG\":\n",
    "        return \"A\"\n",
    "    if aa == \"TAT\" or aa == \"TAC\":\n",
    "        return \"Y\"\n",
    "    if aa == \"TAA\" or aa == \"TAG\" or aa == \"TGA\":\n",
    "        return \"X\"\n",
    "    if aa == \"CAT\" or aa == \"CAC\":\n",
    "        return \"H\"\n",
    "    if aa == \"CAA\" or aa == \"CAG\":\n",
    "        return \"Q\"\n",
    "    if aa == \"AAT\" or aa == \"AAC\":\n",
    "        return \"N\"\n",
    "    if aa == \"AAA\" or aa == \"AAG\":\n",
    "        return \"K\"\n",
    "    if aa == \"GAT\" or aa == \"GAC\":\n",
    "        return \"D\"\n",
    "    if aa == \"GAA\" or aa == \"GAG\":\n",
    "        return \"E\"\n",
    "    if aa == \"TGT\" or aa == \"TGC\":\n",
    "        return \"C\"\n",
    "    if aa == \"TGG\":\n",
    "        return \"W\"\n",
    "    if aa == \"CGT\" or aa == \"CGC\" or aa == \"CGA\" or aa == \"CGG\" or aa == \"AGA\" or aa == \"AGG\":\n",
    "        return \"R\"\n",
    "    if aa == \"GGT\" or aa == \"GGC\" or aa == \"GGA\" or aa == \"GGG\":\n",
    "        return \"G\"\n",
    "    else:\n",
    "        return \"N/A\"\n",
    "    \n",
    "# Start at amino acid position 55 because the amplicon is at that position\n",
    "# Get the codon of of each position by getting 3 consecutive indices\n",
    "def getCodon(position, seq):\n",
    "    try:\n",
    "        index = (position - 55)*3\n",
    "        codon = str(seq[index]) + str(seq[index+1]) + str(seq[index+2])\n",
    "        return codon\n",
    "    except:\n",
    "        return \"N/A\"\n",
    "\n",
    "# Create csv fle to store all the amino acids present at each position\n",
    "csv_file = open('Output_Amplicon.csv', 'w') \n",
    "writer = csv.writer(csv_file)\n",
    "writer.writerow(['Position','AminoAcids', 'Count'])    \n",
    "\n",
    "# Creat a list of positions to find the amino acids\n",
    "list_positions = [90,111,115,122,102,131,142,146,166,168,175]\n",
    "\n",
    "# Add row to the csv by iterating through the fastq file\n",
    "for position in list_positions:\n",
    "    list_AA = []\n",
    "    with open(\"IDT_HM.assembled.fastq\", 'r') as fh:\n",
    "        lines = []\n",
    "        for line in fh:\n",
    "            lines.append(line.rstrip())\n",
    "            # Only choose lines with length 4 which is how fastq file structures\n",
    "            if len(lines) == 4:\n",
    "                # Get the sequence\n",
    "                seq = lines[1]\n",
    "                # Get the codon\n",
    "                codon = getCodon(position,seq)\n",
    "                # Translate the codon\n",
    "                amino_acid = amino_acids(codon)\n",
    "                # If the amino acid not already present then add it to avoid duplicates\n",
    "                if amino_acid not in list_AA:\n",
    "                    list_AA.append(amino_acid)\n",
    "                lines = []\n",
    "    list_AA.remove('N/A')\n",
    "    writer.writerow([position,list_AA,len(list_AA)])\n",
    "                       \n",
    "csv_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0da9d11f-c2f8-487c-bf3b-e190170c0557",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
