{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "662bab61-5824-442f-a921-449837dadd81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2000\n",
      "4000\n",
      "6000\n",
      "8000\n",
      "10000\n",
      "12000\n",
      "14000\n",
      "16000\n",
      "18000\n",
      "20000\n",
      "22000\n",
      "24000\n",
      "26000\n",
      "28000\n",
      "30000\n",
      "32000\n",
      "34000\n",
      "36000\n",
      "38000\n",
      "40000\n",
      "42000\n",
      "44000\n",
      "46000\n",
      "48000\n",
      "50000\n",
      "52000\n",
      "54000\n",
      "56000\n",
      "58000\n",
      "60000\n",
      "62000\n",
      "64000\n",
      "66000\n",
      "68000\n",
      "70000\n",
      "72000\n",
      "74000\n",
      "76000\n",
      "78000\n",
      "80000\n",
      "82000\n",
      "84000\n",
      "86000\n",
      "88000\n",
      "90000\n",
      "92000\n",
      "94000\n",
      "96000\n",
      "98000\n",
      "100000\n",
      "102000\n",
      "104000\n",
      "106000\n",
      "108000\n",
      "110000\n",
      "112000\n",
      "114000\n",
      "116000\n",
      "118000\n",
      "120000\n",
      "122000\n",
      "124000\n",
      "126000\n",
      "128000\n",
      "130000\n",
      "132000\n",
      "134000\n",
      "136000\n",
      "138000\n",
      "140000\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "\"\"\"\n",
    "This is a python script that analyse Amplicon FASTQ file generated with Illumina Sequencing\n",
    "The script classify which amino acids are present at a certain position\n",
    "The purpose is to see which amino acid mutations have been generated\n",
    "All the data are exported into a csv\n",
    "\"\"\"\n",
    "\n",
    "# Translate a codon of 3 bases into the appropriate amino acids\n",
    "def amino_acids(aa):\n",
    "    if aa == \"TTT\" or aa == \"TTC\":\n",
    "        return \"F\"\n",
    "    if aa == \"TTA\" or aa == \"TTG\" or aa == \"CTT\" or aa == \"CTC\" or aa == \"CTA\" or aa == \"CTG\":\n",
    "        return \"L\"\n",
    "    if aa == \"ATT\" or aa == \"ATC\" or aa == \"ATA\":\n",
    "        return \"I\"\n",
    "    if aa == \"ATG\":\n",
    "        return \"M\"\n",
    "    if aa == \"GTT\" or aa == \"GTC\" or aa == \"GTA\" or aa == \"GTG\":\n",
    "        return \"V\"\n",
    "    if aa == \"TCT\" or aa == \"TCC\" or aa == \"TCA\" or aa == \"TCG\" or aa == \"AGT\" or aa == \"AGC\":\n",
    "        return \"S\"\n",
    "    if aa == \"CCT\" or aa == \"CCC\" or aa == \"CCA\" or aa == \"CCG\":\n",
    "        return \"P\"\n",
    "    if aa == \"ACT\" or aa == \"ACC\" or aa == \"ACA\" or aa == \"ACG\":\n",
    "        return \"T\"\n",
    "    if aa == \"GCT\" or aa == \"GCC\" or aa == \"GCA\" or aa == \"GCG\":\n",
    "        return \"A\"\n",
    "    if aa == \"TAT\" or aa == \"TAC\":\n",
    "        return \"Y\"\n",
    "    if aa == \"TAA\" or aa == \"TAG\" or aa == \"TGA\":\n",
    "        return \"X\"\n",
    "    if aa == \"CAT\" or aa == \"CAC\":\n",
    "        return \"H\"\n",
    "    if aa == \"CAA\" or aa == \"CAG\":\n",
    "        return \"Q\"\n",
    "    if aa == \"AAT\" or aa == \"AAC\":\n",
    "        return \"N\"\n",
    "    if aa == \"AAA\" or aa == \"AAG\":\n",
    "        return \"K\"\n",
    "    if aa == \"GAT\" or aa == \"GAC\":\n",
    "        return \"D\"\n",
    "    if aa == \"GAA\" or aa == \"GAG\":\n",
    "        return \"E\"\n",
    "    if aa == \"TGT\" or aa == \"TGC\":\n",
    "        return \"C\"\n",
    "    if aa == \"TGG\":\n",
    "        return \"W\"\n",
    "    if aa == \"CGT\" or aa == \"CGC\" or aa == \"CGA\" or aa == \"CGG\" or aa == \"AGA\" or aa == \"AGG\":\n",
    "        return \"R\"\n",
    "    if aa == \"GGT\" or aa == \"GGC\" or aa == \"GGA\" or aa == \"GGG\":\n",
    "        return \"G\"\n",
    "    else:\n",
    "        return \"N/A\"\n",
    "    \n",
    "# Start at amino acid position 55 because the amplicon is at that position\n",
    "# Get the codon of of each position by getting 3 consecutive indices\n",
    "def getCodon(position, seq):\n",
    "    try:\n",
    "        index = (position - 55)*3\n",
    "        codon = str(seq[index]) + str(seq[index+1]) + str(seq[index+2])\n",
    "        return codon\n",
    "    except:\n",
    "        return \"N/A\"\n",
    "\n",
    "# Get the quality of the codon\n",
    "def getQuality(position, seq):\n",
    "    try:\n",
    "        index = (position - 55)*3\n",
    "        base_qual = [str(seq[index]),str(seq[index+1]),str(seq[index+2])]\n",
    "        return base_qual\n",
    "    except:\n",
    "        return [\"N/A\",\"N/A\",\"N/A\"]\n",
    "    \n",
    "# Compare codon of the original vs the iterated sequence\n",
    "# Return false if they are not the same\n",
    "def compareCodon(position, seq, ori):\n",
    "    ori_codon = getCodon(position,ori)\n",
    "    seq_codon = getCodon(position,seq)\n",
    "    if ori_codon == seq_codon:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "# Create csv fle to store all the amino acids present at each position\n",
    "csv_file = open('Output_ErrorAnalysis.csv', 'w') \n",
    "writer = csv.writer(csv_file)\n",
    "writer.writerow(['No_mutations','Position', 'AminoAcid'])    \n",
    "\n",
    "# Original sequence\n",
    "ori = 'CCAGTGCCCTGGCCTACACTCGTGACCACCCTCACCTACGGAGTCCAGTGCTTCTCCCGGTACCCCGACCACATGAAACAGCACGATTTCTTCAAGTCCGCGATGCCTGAGGGCTATGTTCAGGAACGCACCATATTCTTCAAGGACGACGGTAACTATAAAACCAGGGCTGAAGTTAAGTTCGAGGGGGACACGTTGGTAAATAGAATTGAGTTGAAAGGAATCGATTTCAAAGAAGACGGCAATATTTTGGGCCACAAGCTGGAGTACAACTACAACAGCCACAACGTGTATATTATGGCTGATAAGCAGAAGAACGGCATCAAGGTGAACTTCAAAATCCGACACAATATCGAGGATGGCTCTGTGCAACTGGCCGACCATTACCAGCAAAACACGCCTATCGGCGACGGTCCGGTTTTGCTGCCAGATAACCACTACCTCTCCACGCAATCAGCCCTCTCCAAGGATCC'\n",
    "size_ori = len(ori)\n",
    "\n",
    "# Creat a list of positions to find the amino acids\n",
    "list_positions = [90,111,115,122,102,131,142,146,166,168,175]\n",
    "\n",
    "# Add row to the csv by iterating through the fastq file\n",
    "with open(\"IDT_HM.assembled.fastq\", 'r') as fh:\n",
    "    lines = []\n",
    "    countLine = 0\n",
    "    for line in fh:\n",
    "        lines.append(line.rstrip())\n",
    "        # Only choose lines with length 4 which is how fastq file structures\n",
    "        if len(lines) == 4:\n",
    "            list_diffPos = []\n",
    "            list_mutations = []\n",
    "            count = 0\n",
    "            # Get the sequence\n",
    "            seq = lines[1]\n",
    "            # If indeletion, ignore\n",
    "            if len(seq) != size_ori:\n",
    "                writer.writerow(['indeletion','indeletion','indeletion'])\n",
    "            # Else compare the difference in mutation accross the positions\n",
    "            else:\n",
    "                for position in list_positions:\n",
    "                    if compareCodon(position,seq,ori) == False:\n",
    "                        count = count + 1\n",
    "                        list_diffPos.append(position)\n",
    "                        list_mutations.append(amino_acids(getCodon(position,seq)))\n",
    "                if len(list_mutations) == 0:\n",
    "                    writer.writerow(['WT','WT','WT'])\n",
    "                else:\n",
    "                    writer.writerow([count,list_diffPos,list_mutations])\n",
    "            lines = []\n",
    "            countLine = countLine + 1\n",
    "            if countLine%2000 == 0:\n",
    "                print(countLine)\n",
    "                      \n",
    "csv_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07b060c5-31e7-4029-b29e-8c9212348c37",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
